config-service-impl/src/main/java/org/hypertrace/config/service/ConfigServiceGrpcImpl.java

```diff
@@ -115,21 +128,55 @@ public void getConfig(
   public void getAllConfigs(
       GetAllConfigsRequest request, StreamObserver<GetAllConfigsResponse> responseObserver) {
     try {
-      List<ContextSpecificConfig> contextSpecificConfigList =
+      ConfigResource configResource =
+          new ConfigResource(
+              request.getResourceName(), request.getResourceNamespace(), getTenantId());
+
+      CompletableFuture<Long> totalCountFuture = null;
+
+      // Start count query first (in parallel)
+      if (request.getIncludeTotal()) {
+        totalCountFuture =
+            CompletableFuture.supplyAsync(
+                () -> {
+                  try {
+                    return configStore.getMatchingConfigsCount(configResource, request.getFilter());
+                  } catch (Exception e) {
+                    throw Status.INTERNAL
+                        .withCause(e)
+                        .withDescription("Failed to fetch total count")
+                        .asRuntimeException();
+                  }
+                },
+                configExecutor);
+      }
+
+      // Then run getAllConfigs on the current thread
+      List<ContextSpecificConfig> configList =
           configStore.getAllConfigs(
-              new ConfigResource(
-                  request.getResourceName(), request.getResourceNamespace(), getTenantId()),
+              configResource,
               request.getFilter(),
               request.getPagination(),
               request.getSortByList());
-      responseObserver.onNext(
-          GetAllConfigsResponse.newBuilder()
-              .addAllContextSpecificConfigs(contextSpecificConfigList)
-              .build());
+
+      // Build the response
+      GetAllConfigsResponse.Builder responseBuilder =
+          GetAllConfigsResponse.newBuilder().addAllContextSpecificConfigs(configList);
+
+      if (totalCountFuture != null) {
+        long totalCount = totalCountFuture.join(); // Wait if not finished
+        responseBuilder.setTotalCount(totalCount);
+      }
+
+      responseObserver.onNext(responseBuilder.build());
       responseObserver.onCompleted();
-    } catch (Exception e) {
-      log.error("Get all configs failed for request:{}", request, e);
+    } catch (StatusRuntimeException e) {
+      log.error("Get all configs failed for request: {}", request, e);
       responseObserver.onError(e);
+    } catch (Exception e) {
+      log.error("Get all configs failed for request: {}", request, e);
+      responseObserver.onError(
+          Status.INTERNAL.withCause(e).withDescription("Unexpected error").asRuntimeException());
```

Comment:
nit - can use Status.fromThrowable which will propagate a status exception or wrap it in unknown if needed. Whatever the mechanism, avoiding multiple catch clause would be best.

```suggestion
    } catch (Exception e) {
      log.error("Get all configs failed for request: {}", request, e);
      responseObserver.onError(Status.fromThrowable(e));
    }
```

Reply:
done